client/blocks/user-mentions/with-user-mentions.jsx

```diff
@@ -18,33 +22,252 @@ import React from 'react';
+				return;
+			}
+
+			let nextIndex;
+
+			// Cancel the cursor move.
+			event.preventDefault();
+
+			if ( event.keyCode === keys.downArrow ) {
+				nextIndex = ( selectedIndex + 1 ) % this.matchingSuggestions.length;
+			} else {
+				nextIndex = selectedIndex - 1;
+
+				if ( nextIndex < 0 ) {
+					nextIndex = this.matchingSuggestions.length - 1;
+				}
+			}
+
+			this.setState( { selectedSuggestionId: this.matchingSuggestions[ nextIndex ].ID } );
+		};
+
+		handleKeyUp = event => {
+			if ( includes( [ keys.downArrow, keys.upArrow ], event.keyCode ) ) {
+				return;
+			}
+
+			if ( includes( [ keys.spaceBar, keys.esc ], event.keyCode ) ) {
+				return this.hidePopover();
+			}
+
+			if ( includes( [ keys.enter ], event.keyCode ) ) {
+				if ( ! this.state.showPopover || this.matchingSuggestions.length === 0 ) {
+					return;
+				}
+
+				event.preventDefault();
+
+				const suggestion = this.getSuggestion();
+
+				if ( suggestion ) {
+					this.insertSuggestion( suggestion );
+				}
+
+				return this.hidePopover();
+			}
+
+			const query = this.getQueryText();
+
+			this.setState( {
+				showPopover: query !== null,
+				selectedSuggestionId: null,
+				query,
+			} );
+		};
+
+		getQueryText() {
+			const node = this.textInput.current;
+			// (?:^|\\s) means start of input or whitespace
+			// ([A-Za-z0-9_\+\-]*) means 0 or more instances of: A-Z a-z 0-9 _ + -
+			const matcher = new RegExp( '(?:^|\\s)@([A-Za-z0-9_+-]*)$', 'gi' );
+			const textBeforeCaret = node.value.slice( 0, node.selectionEnd );
+			const match = matcher.exec( textBeforeCaret );
+
+			return match && match.length > 1 ? match[ 1 ] : null;
+		}
+
+		getPosition() {
+			const node = this.textInput.current;
+			const nodeRect = node.getBoundingClientRect();
+			const caretPosition = getCaretCoordinates( node, node.selectionEnd );
+			const position = {
+				left: nodeRect.left + caretPosition.left,
+				top: nodeRect.top + caretPosition.top + 28,
+			};
+
+			console.log( position ); // eslint-disable-line no-console
+
+			return position;
+		}
+
+		getSuggestion() {
+			const index = this.getSelectedSuggestionIndex();
+
+			return index > -1 ? this.matchingSuggestions[ index ] : null;
+		}
+
+		getSelectedSuggestionIndex() {
+			if ( ! this.state.selectedSuggestionId ) {
+				return 0;
+			}
+
+			return findIndex(
+				this.matchingSuggestions,
+				( { ID: id } ) => id === this.state.selectedSuggestionId
+			);
+		}
+
+		getMatchingSuggestions( suggestions, query ) {
+			if ( query ) {
+				query = escapeRegExp( query );
+				const matcher = new RegExp( `^${ query }|\\s${ query }`, 'ig' ); // Start of string or preceded by a space.
+
+				suggestions = suggestions.filter( ( { user_login: login, display_name: name } ) =>
+					matcher.test( `${ login } ${ name }` )
+				);
+			}
+
+			return suggestions.slice( 0, 10 );
+		}
+
+		insertSuggestion = ( { user_login: userLogin } ) => {
```

Comment:
Maybe `autocompleteSuggestion` would be easier to understand?

Reply:
I agree, but will stick with `insertSuggestion` for the moment to be consistent with Editor mentions.